ducktape/cluster/vagrant.py

```diff
@@ -24,10 +28,53 @@ class VagrantCluster(JsonCluster):
     advertised, this assumes that the Vagrant VM's name is a routeable hostname on all the hosts.
     """
 
-    def __init__(self):
-        hostname, ssh_hostname, username, flags = None, None, None, ""
-        nodes = []
+    def __init__(self, *args, **kwargs):
+        is_read_from_file = False
 
+        cluster_file = kwargs.get("cluster_file")
+        if cluster_file is not None:
+            try:
+                cluster_json = json.load(open(os.path.abspath(cluster_file)))
+                is_read_from_file = True
+            except IOError:
+                # It is OK if file is not found. Call vagrant ssh-info to read the cluster info.
+                pass
+
+        if not is_read_from_file:
+            cluster_json = {
+                "nodes": self._get_nodes_from_vagrant()
+            }
+
+        super(VagrantCluster, self).__init__(cluster_json)
+
+        # Load externally_routable_ip for each account
+        if is_read_from_file:
+            externally_routable_ip = cluster_json["externally_routable_ip"]
+            for node_account in self.available_nodes:
+                node_account.externally_routable_ip = externally_routable_ip[str(node_account)]
+        else:
+            # go through and find fully qualified domain name for each node
+            # this makes it possible to not require write access to /etc/hosts on the test driver machine
+            is_aws = self._is_aws()
+
+            externally_routable_ip = {}
+            for node_account in self.available_nodes:
+                node_account.externally_routable_ip = self._externally_routable_ip(is_aws, node_account)
+                externally_routable_ip[str(node_account)] = node_account.externally_routable_ip
+
+        # If cluster_file is specified, cache cluster to the file
+        if cluster_file is not None:
+            cluster_json["externally_routable_ip"] = externally_routable_ip
+            try:
+                json.dump(cluster_json, open(cluster_file, "w"))
+            except IOError:
+                # Do not create a new file and cache cluster info in unit test
```

Comment:
I am very sure that I added this try/catch because one unit test failed earlier. But I couldn't reproduce this now. Maybe the behavior changed due to some other commits checked-in between.

I will remove this try/catch.